trlx/pipeline/offline_pipeline.py

```diff
@@ -14,45 +19,53 @@
 from trlx.pipeline import BasePipeline, BaseRolloutStore, register_datapipeline
 
 
-def tokenize_dialogue(dialogue: Union[str, List[str]], tokenizer, max_length=2048) -> List[int]:  # noqa: C901
+@dataclass
+class DialogMessage:
+    is_output: bool
+    tokens: Tuple[int]
+
+
+def tokenize_dialogue(  # noqa: C901
+    dialogue: Union[str, List[str]], tokenizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast], max_length=2048
+) -> List[DialogMessage]:
     """
     Tokenize sample with the interleaved form of (prompt_1, output_1, prompt_2, output_2...)
     """
     if isinstance(dialogue, str):
         dialogue = [tokenizer.bos_token, dialogue]
     elif isinstance(dialogue, tuple):
+        if len(dialogue) % 2 != 0:
+            raise ValueError("Dialogue must have an even number of phrases, alternating prompt and output")
         dialogue = list(dialogue)
 
-    out = []
-    ctx_length = max_length - 1
+    if not dialogue[-1].endswith(tokenizer.eos_token):
+        dialogue[-1] = dialogue[-1] + tokenizer.eos_token
```

Comment:
If samples are truncated down the line, `<eos>` will be trimmed as well, so that's not the same behavior as on the main. Duy, in his experiments, said that this is required for T5

Reply:
https://wandb.ai/uwu1/trlx/reports/t5-truncation-change--VmlldzozOTAwNzYw
It does seem to change it but only for the later training iterations